### PR TITLE
Automate: Verify if APIBinding binds with exported custom resources

### PR DESCRIPTION
**Pull request by zxiao-redhat**
This is a test case proposed under OCP-59375 that verifies if APIBinding works with exported custom resources.

Test: 
```
passed: (1m18s) 2023-02-02T14:26:55 "[area/apiexports] Author:zxiao-Critical-[KCP] Verify if APIBinding binds with exported custom resource [Suite:kcp/smoke/parallel]"
```

### CODE REVIEW

File: test/extended/util/client.go [hunk 1]
```diff
@@ -455,7 +455,7 @@ func (c *CLI) SetupWorkSpaceWithNamespace() {
 func (c *CLI) SetupWorkSpaceWithSpecificPath(serverURL string) {
 	newWorkSpace := names.SimpleNameGenerator.GenerateName(fmt.Sprintf("e2e-test-%s-", c.kubeFramework.BaseName))
 	e2e.Logf("Creating workspace %q", newWorkSpace)
-	output, errinfo := c.WithoutKubeconf().WithoutWorkSpaceServer().Run("ws").Args("create", "--server="+serverURL, newWorkSpace).Output()
+	output, errinfo := c.WithoutKubeconf().WithoutWorkSpaceServer().Run("kcp").Args("ws", "create", "--server="+serverURL, newWorkSpace).Output()
```
Comment (kasturinarra) [inline test/extended/util/client.go]: @zimo-xiao do not think this change is required, in your system sym links might have been broken and that might be the reason you might be getting an error.

Comment (kasturinarra) [reply test/extended/util/client.go]: something like below
ln -sf /home/kubectl-workspace /home/knarra/kubectl-ws
ln -sf /home/kubectl-workspace /home/knarra/kubectl-workspaces

File: test/extended/testdata/apibinding/role_hack.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+metadata:
+  name: bind-apiexport-spi
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: bind-apiexport-spi
+subjects:
+- apiGroup: rbac.authorization.k8s.io
+  kind: User
+  name: "$(USERNAME)"
```
Comment (kasturinarra) [inline test/extended/testdata/apibinding/role_hack.yaml]: @zxiao-redhat any reason you are changing this for a particular user ? we could may be do it for all ?

File: test/extended/util/client.go (hunk 1 above)
Comment (zxiao-redhat) [reply test/extended/util/client.go]: I agree with you and fixed my local link. I had reverted the change.

File: test/extended/testdata/apibinding/role_hack.yaml (hunk 2 above)
Comment (zxiao-redhat) [reply test/extended/testdata/apibinding/role_hack.yaml]: @kasturinarra According to my research, ClusterRoleBinding only applies in a specific context under some namespace or user. Therefore, it is not possible to grant permission to all users.

Comment (kasturinarra) [reply test/extended/testdata/apibinding/role_hack.yaml]: @zimo-xiao can you try and see if below works ? I have been following this way
```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: system:kcp:authenticated:apiexport:tenancy:bind
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: system:kcp:apiexport:tenancy:bind
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: system:authenticated
```

Comment (zxiao-redhat) [reply test/extended/testdata/apibinding/role_hack.yaml]: Hi @kasturinarra, thank you for the suggestion. I had just used you method and updated the repo. PTAL. Thanks!
Test: passed: (45.1s) 2023-02-09T10:45:38 "[area/apiexports] Author:zxiao-Critical-[KCP] Verify if APIBinding binds with exported custom resource [Suite:kcp/smoke/parallel]"

File: test/extended/apibinding/apibinding.go [hunk 3]
```diff
@@ -43,4 +46,52 @@ var _ = g.Describe("[area/apiexports]", func() {
...
+		_, err = exutil.CreateResourceFromTemplate(k, exportTemplate)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// THIS IS A HACK: extract username from permission error
+		g.By("# HACK: get username for bug fix")
+		output, _ := k.Run("get").Args("clusterworkspace").Output()
+		usernameRegex := regexp.MustCompile(`User "(.*?)" cannot list resource`)
+		usernameMatch := usernameRegex.FindStringSubmatch(output)
+		o.Expect(usernameMatch).NotTo(o.BeEmpty())
+		username := usernameMatch[1]
```
Comment (kasturinarra) [inline test/extended/apibinding/apibinding.go]: @zimo-xiao do we still need the above code ?

File: test/extended/apibinding/apibinding.go [hunk 4]
```diff
@@ -43,4 +46,52 @@ var _ = g.Describe("[area/apiexports]", func() {
...
+
+		// BUG: https://github.com/kcp-dev/kcp/issues/1939
+		g.By("# BUG: apply role binding hack to allow api-binding for non-admin user")
+		roleHackTemplate := exutil.FixturePath("testdata", "apibinding", "role_hack.yaml")
+		_, err = exutil.CreateResourceFromTemplateWithVariables(k, roleHackTemplate, map[string]string{
+			"USERNAME": username,
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
```
Comment (kasturinarra) [inline test/extended/apibinding/apibinding.go:65]: @zimo-xiao also this code

File: test/extended/testdata/apibinding/role_hack.yaml (hunk 2 above)
Comment (kasturinarra) [reply test/extended/testdata/apibinding/role_hack.yaml]: @zimo-xiao also do not see the updated fies or am i missing something ? could you please help check ? thanks !!

Comment (zxiao-redhat) [reply test/extended/testdata/apibinding/role_hack.yaml]: Hi @kasturinarra I apologize. 😂 I committed my changes to another branch. I just updated my code base, please take a look.

Test: passed: (26.6s) 2023-02-12T14:05:14 "[area/apiexports] Author:zxiao-Critical-[KCP] Verify if APIBinding binds with exported custom resource [Suite:kcp/smoke/parallel]"

File: test/extended/apibinding/apibinding.go [hunk 5]
```diff
@@ -43,4 +43,40 @@ var _ = g.Describe("[area/apiexports]", func() {
...
+	g.It("Author:zxiao-Critical-[KCP] Verify if APIBinding binds with exported custom resource", func() {
+		g.By("# Create a test workspace")
+		k.SetupWorkSpace()
+
+		g.By("# Create custom resource cowboy using APIResourceSchema")
+		rsTemplate := exutil.FixturePath("testdata", "apibinding", "api_rs.yaml")
+		_, err := exutil.CreateResourceFromTemplate(k, rsTemplate)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("# Export custom resource using APIExport")
```
Comment (kasturinarra) [inline test/extended/apibinding/apibinding.go]: @zxiao-redhat should we simply say "create apiexport for custom resource " ?

File: test/extended/apibinding/apibinding.go [hunk 6]
```diff
@@ -43,4 +43,40 @@ var _ = g.Describe("[area/apiexports]", func() {
 		o.Expect(err).NotTo(o.HaveOccurred())
 		o.Expect(finalizers).Should(o.ContainSubstring("workload.kcp.dev/syncer"))
 	})
+
+	g.It("Author:zxiao-Critical-[KCP] Verify if APIBinding binds with exported custom resource", func() {
+		g.By("# Create a test workspace")
+		k.SetupWorkSpace()
+
+		g.By("# Create custom resource cowboy using APIResourceSchema")
```
Comment (kasturinarra) [inline test/extended/apibinding/apibinding.go]: @zimo-xiao should we say "create cowboy custom resource APIResourceSchema"

File: test/extended/apibinding/apibinding.go [hunk 7]
```diff
@@ -43,4 +43,40 @@ var _ = g.Describe("[area/apiexports]", func() {
...
+		_, err = exutil.CreateResourceFromTemplate(k, exportTemplate)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// BUG: https://github.com/kcp-dev/kcp/issues/1939
+		g.By("# BUG: apply role binding hack to allow api-binding for non-admin user")
+		roleHackTemplate := exutil.FixturePath("testdata", "apibinding", "role_hack.yaml")
+		_, err = exutil.CreateResourceFromTemplate(k, roleHackTemplate)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("# Bind to custom resource using APIBinding")
```
Comment (kasturinarra) [inline test/extended/apibinding/apibinding.go]: @zxiao-redhat say "create apibinding for custom resource cowboy " ?

**Pull request merged**